### DISCUSSION TITLE
Is there a way to write a rule containing "OR" in the Matcher ?

### DISCUSSION THREAD

**Discussion by lio-data** [explosion/spaCy#12818] [Help: Other Questions] [answered]
Hi,
I am using the Matcher to mark some text. I have quite a few complex rules and I would need to generalize it.  Here is one of them :
[[
      {"IS_SENT_START": true}, {"IS_SENT_START": false, "OP": "{0,9}"}, {"LOWER": ":"},
      {"OP": "?"}, {"LIKE_NUM": true}, {"_": {"is_unit": true}},  {"OP": "?"},{"TEXT": {"REGEX": "\\n"}}
]]

I would want to have the Matcher use LIKE_NUM or another custom attribute "._.is_cn". is_cn is simply a custom extension that I put on complicated text describing numbers likes 64x7 or 7/2. My pipeline recognize these complex representation of numbers.

So to some extent I would like the Matcher to be able to use this pattern : 
[[
      {"IS_SENT_START": true}, {"IS_SENT_START": false, "OP": "{0,9}"}, {"LOWER": ":"},
      {"OP": "?"}, {"LIKE_NUM": true} OR  {"_": {"is_cn": true}}, {"_": {"is_unit": true}},  {"OP": "?"},{"TEXT": {"REGEX": "\\n"}}
]]

I have not seen any way to do that... 

Currently the only I see to achieve my goal is either 
- to duplicate all my patterns and replace the {"LIKE_NUM": true} by  {"_": {"is_cn": true}} but this is just not great... my list of patterns will become unreadable ! 
- to create another custom boolean token extension which I put as False unless LIKE_NUM or is_cn is True and then use that extension for the Matcher. This is currently my favorite option but it is not great either. LIKE_NUM is not an writable attribute of token as otherwise I would have simply put LIKE_NUM = True for all token where ._.is_cn = True

Any better idea ?

**Comment by svlandeg** [comment, marked as answer]
Hey @lio-data,

This is perhaps a naive suggestion, but when you're assigning things to be `._.is_cn` in your custom code, can't you also label all numbers as such, so that you don't need the `LIKE_NUM` part anymore? For English, you can find its definition [here](https://github.com/explosion/spaCy/blob/master/spacy/lang/en/lex_attrs.py).

One other alternative that is maybe a little more involved, but would result in nicer matcher patterns is this: you can influence the definition of the  `LIKE_NUM` attribute directly by going in and writing [new language defaults](https://github.com/explosion/spaCy/blob/master/spacy/lang/en/__init__.py) and replacing the `lex_attr_getters ` with your own implementation. For more information on that, see here: https://spacy.io/usage/linguistic-features#language-data

**Reply by lio-data** [reply]
@svlandeg thank you for the reply.  Your suggestion is in fact maybe the best option. I am considering using `._.is_cn` to mark all kind of numbers.  My other alternative is to create a function/class to create the rules for the matcher. Thank you !
